### ISSUE TITLE
remove NullableArrays

### ISSUE THREAD

**Issue by shashi** [JuliaComputing/TextParse.jl#18]
Just use `Vector{Nullable{T}}` for now.

see https://github.com/JuliaStats/NullableArrays.jl/pull/194

**Comment by quinnj** [comment]
There's also https://github.com/JuliaData/Nulls.jl

**Comment by davidanthoff** [comment]
Another options is [DataValues.jl](https://github.com/davidanthoff/DataValues.jl) and [DataValueArrays.jl](https://github.com/davidanthoff/DataValueArrays.jl) (not yet registered, but soon).

These are essentially slightly modified versions of ``Nullable`` and ``NullableArray`` that I use for all missing values in [IterableTables.jl](https://github.com/davidanthoff/IterableTables.jl) and [Query.jl](https://github.com/davidanthoff/Query.jl). One of the modifications is that all comparison operators return ``Bool`` always and never ``Nullable{Bool}``. The array version also should have the normal array broadcasting behavior etc. and not the automatic lifting versions that you get for ``NullableArray``.

If you are using ``Vector{Nullable{T}}`` things will break as soon as anyone is loading ``NullableArrays`` because of the type piracy in that package. Essentially you'll get the ``Nullable{Bool}`` return type for ``==`` on ``Nullable``s as soon as anyone loads that package. So even just using ``Vector{DataValue{T}}`` should make for a more stable story.

Having said all of that, the perfect solution of course would be if one could specify what container one wanted to use for missing values, so that this code would work for ``NullableArray``, ``DataValueArray``, ``Vector{Nullable{T}}``, ``Vector{DataValue{T}}`` and ``Vector{Union{T,Null}}``.

**Comment by shashi** [comment]
> If you are using Vector{Nullable{T}} things will break as soon as anyone is loading NullableArrays because of the type piracy in that package.

Aw man.

> the perfect solution of course would be if one could specify what container one wanted to use for missing values

I think this can be arranged, but we should pick some default.

**Comment by shashi** [comment]
From what I can understand `Nulls.jl` needs the compiler optimizations to become readily available. I love the stuff so far in the test file! Would be good if there's some documentation so that this package / JuliaDB can point to it.

DataValues.jl seems to offer basically similar API. Pretty nice! Probably a good default to start with, the other contender for the default being Nullable itself.

**Comment by davidanthoff** [comment]
The problem with any of the ``Vector{T}`` (whether ``T`` is ``Nullable``, ``DataValue`` or ``Union{T,Null}``) is that you would have top copy all values of any column that has been read already at the promote step that kicks in once you detect a missing value, right? As far as I can see, only ``NullableArray`` and ``DataValueArray`` would currently allow you to reuse the array with the values read so far inside an array type that also supports missingness, right? That might make quite a difference in terms of performance.

**Comment by shashi** [comment]
That's right. It could save a couple of seconds.

**Comment by davidanthoff** [comment]
Alright, I'm trying to get [DataValueArrays.jl](https://github.com/davidanthoff/DataValueArrays.jl) into shape now. And if you have any feedback on it, please let me know. I would want that package to work exactly in the kind of scenario discussed here, so happy to make changes and accommodate requirements here.

**Comment by shashi** [comment]
Sounds good, thanks!

**Comment by shashi** [comment]
It looks like I can swap out `DataValue{T}` for `T?` once it becomes efficient out of the box. So I'll go with DataValue and DataValueArrays.jl for now. If `const null = DataValue()` was exported, the user API would be exactly the same...

**Comment by davidanthoff** [comment]
I'm in the process of moving ``DataValueArray`` into DataValues.jl. Makes it easier to get the broadcast story right. I'll try and tag a version tonight. Current state of affairs is at https://github.com/davidanthoff/DataValues.jl/pull/21.

**Comment by davidanthoff** [comment]
Oh, and ``DataValues.jl`` exports ``const NA = DataValue{Union{}}()``.

**Comment by shashi** [comment]
Sounds great! Let's do this!

**Comment by davidanthoff** [comment]
https://github.com/JuliaLang/METADATA.jl/pull/11229 has ``DataValueVector`` in the normal DataValues.jl package.

**Comment by shashi** [comment]
Doesn't matter to me much, but I thought this should work
```julia
julia> x[1] = Nullable(3)
ERROR: MethodError: Cannot `convert` an object of type Nullable{Int64} to an object of type Int64
This may have arisen from a call to the constructor Int64(...),
since type constructors fall back to convert methods.
Stacktrace:
 [1] setindex!(::DataValues.DataValueArray{Int64,1}, ::Nullable{Int64}, ::Int64) at /home/shashi/.julia/v0.6/DataValues/src/array/indexing.jl:76
```

**Comment by davidanthoff** [comment]
https://github.com/davidanthoff/DataValues.jl/issues/23

**Comment by nalimilan** [comment]
> The problem with any of the Vector{T} (whether T is Nullable, DataValue or Union{T,Null}) is that you would have top copy all values of any column that has been read already at the promote step that kicks in once you detect a missing value, right? As far as I can see, only NullableArray and DataValueArray would currently allow you to reuse the array with the values read so far inside an array type that also supports missingness, right? That might make quite a difference in terms of performance.

That's not right. You just need to start with an `Array{Union{T, Null}}` to be ready to handle the possibility of a missing values. That's essentially what you do when allocating a `NullableArray`/`DataValueArray`. With  `Array{Union{T, Null}}` on Julia 0.7, calling `convert(Array{T}, x::Array{Union{T, Null}})` is very efficient and doesn't make any copy.

**Comment by shashi** [comment]
What about `convert(Array{Union{T, Null}}, x::Array{T})`? I think that's what David is talking about. This package does this when a nullable value is found for the first time in a column. I should say this is not a huge concern. These promotions usually account for < 10% of the time and memory is quickly gc'd.

**Comment by nalimilan** [comment]
Ah, OK. In theory, given the memory layout of `Array{Union{T, Null}}`, no copy should be necessary, but this optimization doesn't seem to be implemented in 0.7. Would be worth filing an issue if you consider using that design at some point.
